### ISSUE TITLE
Unix struct copying issue 

### ISSUE THREAD

**Issue by janvorli** [dotnet/coreclr#6585]
The change #6037 caused a regression that was discovered by @mellinoe when trying to run csc with the latest coreclr. Commit e06211d164f565242d1a8fab9d88965210eb7f64 works fine while commit 68d5832f204e791943a7e1c5a7f0c80063cf3c7a for #6037 causes a segmentation fault in the test.
The issue is caused by a structure that's a member of another structure being copied to stack as a parameter for calling a method from a wrong offset (zero instead of the real offset of the sub-structure).
It happens when Microsoft.CodeAnalysis.CSharp.OverloadResolution.BinaryOperatorOverloadResolution calls Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterOperator. The function first gets 
the Microsoft.CodeAnalysis.CSharp.BinaryOperatorAnalysisResult to rbp-0x2b8:
```ASM
00007fff7f213a5e 488db548fdffff       lea     rsi, [rbp - 0x2b8]
00007fff7f213a65 8b55d0               mov     edx, dword ptr [rbp - 0x30]
00007fff7f213a68 393f                 cmp     dword ptr [rdi], edi
00007fff7f213a6a e8e1e1fdff           call    0x7fff7f1f1c50 (System.Collections.Immutable.ImmutableArray`1+Builder[[Microsoft.CodeAnalysis.CSharp.BinaryOperatorAnalysisResult, Microsoft.CodeAnalysis.CSharp]].get_Item(Int32), mdToken: 00000000060003A7)
```
And it is followed by the call to the Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterOperator and prepares the stack parameters as follows (see the added comments starting with <<<):
```ASM
00007fff7f213a6f 488d3c24             lea     rdi, [rsp]
00007fff7f213a73 488db5c0fdffff       lea     rsi, [rbp - 0x240]
00007fff7f213a7a 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213a7d 48890c24             mov     qword ptr [rsp], rcx
00007fff7f213a81 4883c608             add     rsi, 0x8
00007fff7f213a85 4883c708             add     rdi, 0x8
00007fff7f213a89 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213a8c 48894c2408           mov     qword ptr [rsp + 0x8], rcx
00007fff7f213a91 4883c608             add     rsi, 0x8
00007fff7f213a95 4883c708             add     rdi, 0x8
00007fff7f213a99 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213a9c 48894c2410           mov     qword ptr [rsp + 0x10], rcx
00007fff7f213aa1 4883c608             add     rsi, 0x8
00007fff7f213aa5 4883c708             add     rdi, 0x8
00007fff7f213aa9 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213aac 48894c2418           mov     qword ptr [rsp + 0x18], rcx
00007fff7f213ab1 4883c608             add     rsi, 0x8
00007fff7f213ab5 4883c708             add     rdi, 0x8
00007fff7f213ab9 48a5                 movsq   qword ptr es:[rdi], qword ptr [rsi]
00007fff7f213abb 48a5                 movsq   qword ptr es:[rdi], qword ptr [rsi]
<<< The address below will be referred to as RBP+0x40 by the callee and is expected to BinaryOperatorSignature
00007fff7f213abd 488d7c2430           lea     rdi, [rsp + 0x30]  
<<< The address below is where the BinaryOperatorAnalysisResult was just fetched
00007fff7f213ac2 488db548fdffff       lea     rsi, [rbp - 0x2b8] 
<<< Now we copy the first 48 bytes of the BinaryOperatorAnalysisResult instead of the BinaryOperatorSignature which is at offset 0x30 in the BinaryOperatorAnalysisResult!
00007fff7f213ac9 488b0e               mov     rcx, qword ptr [rsi]            
00007fff7f213acc 48894c2430           mov     qword ptr [rsp + 0x30], rcx
00007fff7f213ad1 4883c608             add     rsi, 0x8
00007fff7f213ad5 4883c708             add     rdi, 0x8
00007fff7f213ad9 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213adc 48894c2438           mov     qword ptr [rsp + 0x38], rcx
00007fff7f213ae1 4883c608             add     rsi, 0x8
00007fff7f213ae5 4883c708             add     rdi, 0x8
00007fff7f213ae9 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213aec 48894c2440           mov     qword ptr [rsp + 0x40], rcx
00007fff7f213af1 4883c608             add     rsi, 0x8
00007fff7f213af5 4883c708             add     rdi, 0x8
00007fff7f213af9 488b0e               mov     rcx, qword ptr [rsi]
00007fff7f213afc 48894c2448           mov     qword ptr [rsp + 0x48], rcx
00007fff7f213b01 4883c608             add     rsi, 0x8
00007fff7f213b05 4883c708             add     rdi, 0x8
00007fff7f213b09 48a5                 movsq   qword ptr es:[rdi], qword ptr [rsi]
00007fff7f213b0b 48a5                 movsq   qword ptr es:[rdi], qword ptr [rsi]
00007fff7f213b0d 488bfb               mov     rdi, rbx
00007fff7f213b10 498bf6               mov     rsi, r14
00007fff7f213b13 498bd7               mov     rdx, r15
00007fff7f213b16 498bcc               mov     rcx, r12
00007fff7f213b19 e8e27020ff           call    0x7fff7e41ac00 (Microsoft.CodeAnalysis.CSharp.OverloadResolution.BetterOperator(Microsoft.CodeAnalysis.CSharp.BinaryOperatorSignature, Microsoft.CodeAnalysis.CSharp.BinaryOperatorSignature, Microsoft.CodeAnalysis.CSharp.BoundExpression, Microsoft.CodeAnalysis.CSharp.BoundExpression, System.Collections.Generic.HashSet`1<Microsoft.CodeAnalysis.DiagnosticInfo> ByRef), mdToken: 00000000060004D9)
```

**Comment by janvorli** [comment]
Please see the jitdumps for both of the methods here:
https://gist.github.com/janvorli/b35ffd8916acc5bd1c79c7e416284e39

**Comment by janvorli** [comment]
cc: @dotnet/jit-contrib

**Comment by RussKeldorph** [comment]
@BruceForstall Could you take a look and revert #6037 if that is the most expedient way to fix this short term?

**Comment by CarolEidt** [comment]
I actually think this may be the same issue as #6222. I fixed an issue with very similar symptoms on arm, in my "Less Conservative GtObj" changes. I was originally hoping it was the same bug, but the issue was in arm-only code. However, since the issues seemed so similar, I was looking at the method that had the issue there.
In `TypeLoadException:SetMessageField()`, there is a call to `String.FormatHelper()`, and it is passing a struct with 4 ref fields.
On x64/ux it is using the base address of the source struct to load each field (so it is loading the same field 4 times).
@RussKeldorph @BruceForstall - I will be away this afternoon, but can take a look at this later if you'd like.

**Comment by mellinoe** [comment]
@janvorli Do you know if we could pull out a simple regression test from the `Microsoft.CodeAnalysis.CSharp.OverloadResolution.BinaryOperatorOverloadResolution` call chain? Or is it too entangled with the rest of the compiler code to pull out?

**Comment by janvorli** [comment]
It seems worth trying to create a simple repro that would mimic just those two functions, but as a standalone thing. What I mean is to have two structs, one having several object members, the other having several members at the beginning, then the other struct. And using it the way the two funcs from the compiler do. Have a method that generates the larger struct and calls another method that accepts the smaller struct, passing it the sub-struct from the larger one by value. The second method would try to do something with one of the object members of the passed in struct. 
If we are lucky, we might hit the issue. But it is possible that the issue depends on other factors.

**Comment by BruceForstall** [comment]
@CarolEidt Please take a look. If you don't find anything quickly, we should consider reverting the original change.

**Comment by briansull** [comment]
These codesize improvement from  #6037 look suspicious to me:
In particular the improvement in Microsoft.CodeAnalysis.CSharp.dasm
 seems more likely due to failures in the Jitting of methods rather than improved codegen.

Total bytes of diff: -204311
    diff is an improvement.
Top file regressions by size (bytes):
          11 : System.Text.RegularExpressions.dasm
Top file improvements by size (bytes):
     -157369 : Microsoft.CodeAnalysis.CSharp.dasm
      -21422 : Microsoft.CodeAnalysis.dasm
      -12614 : System.Linq.Parallel.dasm
       -7369 : System.Private.CoreLib.dasm
       -3807 : System.Collections.Immutable.dasm
17 total files with size differences.

**Comment by CarolEidt** [comment]
I believe I understand the problem, and have a tentative fix. I'm now working on getting a small repro validate it. I believe that the problem is pretty straightforward, in that the source of the `GT_OBJ` (struct argument) node was a lclFldAddr, and codegen was treating it as a lclVarAddr, i.e. not adding in the offset. The bug is in `genConsumePutStructArgStk()` and I believe it should apply to both the case where the struct is too large to unroll, or where it has gc refs. I have a repro for the latter, but not (yet) the former.

**Comment by CarolEidt** [comment]
I have a PR #6616 that I believe fixes this issue, along with a test that repro'd the behavior.

**Comment by CarolEidt** [comment]
Closed with #6616
